File: epistimi-frontend/src/dto/student-grades.ts
import { ClassificationGradeTeacherResponse, ClassificationGradeValueResponse } from './classification-grade';
import { GradeResponse } from './grade';

export interface StudentsGradesResponse {
  students: StudentGradesResponse[];
}

export interface StudentGradesResponse {
  id: string;
  firstName: string;
  lastName: string;
  subjects: StudentGradesSubjectResponse[];
}

export interface StudentGradesSubjectResponse {
  id: string;
  name: string;
  firstSemester: StudentGradesSubjectSemesterResponse;
  secondSemester: StudentGradesSubjectSemesterResponse;
  average: StudentGradesAverageResponse;
  classification: StudentSubjectClassificationResponse;
}

export interface StudentSubjectClassificationResponse {
  proposal?: StudentSubjectClassificationGradeResponse;
  final?: StudentSubjectClassificationGradeResponse;
}

export interface StudentSubjectClassificationGradeResponse {
  id: string;
  issuedBy: ClassificationGradeTeacherResponse;
  issuedAt: Date;
  updatedAt?: Date;
  value: ClassificationGradeValueResponse;
}

export interface StudentGradesSubjectSemesterResponse {
  grades: GradeResponse[];
  average: StudentGradesAverageResponse;
  classification: StudentSubjectClassificationResponse;
}

export interface StudentGradesAverageResponse {
  student?: string;
  subject?: string;
}

File: epistimi-backend/src/main/kotlin/pl/edu/wat/wcy/epistimi/grade/domain/StudentGrades.kt
package pl.edu.wat.wcy.epistimi.grade.domain

import pl.edu.wat.wcy.epistimi.student.domain.StudentId
import pl.edu.wat.wcy.epistimi.subject.domain.SubjectId
import java.math.BigDecimal

data class StudentsGrades(
    val students: List<StudentGrades>,
)

data class StudentGrades(
    val id: StudentId,
    val firstName: String,
    val lastName: String,
    val subjects: List<StudentSubjectGradesSummary>,
)

data class StudentSubjectGradesSummary(
    val id: SubjectId,
    val name: String,
    val firstSemester: StudentSubjectSemesterGradesSummary,
    val secondSemester: StudentSubjectSemesterGradesSummary,
    val average: StudentGradesAverage,
    val classification: StudentSubjectClassification,
)

data class StudentSubjectClassification(
    val proposal: ClassificationGrade?,
    val final: ClassificationGrade?,
)

data class StudentSubjectSemesterGradesSummary(
    val grades: List<Grade>,
    val average: StudentGradesAverage,
    val classification: StudentSubjectClassification,
)

data class StudentGradesAverage(
    val student: BigDecimal?,
    val subject: BigDecimal?,
)
